## 과제 1
ReLu activation function과 derivative function을 구현해보세요
- Hint : np.maximum 함수 사용하면 편리합니다
- 다른 방법 사용하셔도 무방합니다


In [1]:
import numpy as np

In [2]:
def relu(x):
 return np.maximum(0,x)

In [3]:
def d_relu(x):
  if x>0:
    val = 1
  else:
    val = 0
  return val

## 과제 2
Deep Learning Basic 코드 파일의 MLP implementation with Numpy library using MNIST dataset 코드 참고해서
Three layer MLP 일 때의 backward_pass 함수를 완성해주세요.   
- Hint : 코드 파일의 예시는 Two layer MLP


In [4]:
def backward_pass(x, y_true, params):

  dS3 = params["A3"] - y_true

  grads = {}

  grads["dW3"] = np.dot(dS3, params["A2"].T)/x.shape[1]
  grads["db3"] = (1/x.shape[1])*np.sum(dS3, axis=1, keepdims=True)/x.shape[1]

  dA2 = np.dot(params["W3"].T, dS3)
  dS2 = dA2 * d_relu(params["S2"])

  grads["dW2"] = np.dot(dS2, params["A1"].T)/x.shape[1]
  grads["db2"] = np.sum(dS2, axis=1, keepdims=True)/x.shape[1]

  dA1 = np.dot(params["W2"].T, dS2)
  dS1 = dA1 * d_sigmoid(params["S1"])

  grads["dW1"] = np.dot(dS1, x.T)/x.shape[1]
  grads["db1"] = np.sum(dS1, axis=1, keepdims=True)/x.shape[1]

  return grads

## 과제 3
Deep Learning Basic 코드 파일의 MLP implementation with Pytorch library using MNIST dataset 코드 참고해서
Three layer MLP를 구한후, 학습을 돌려 보세요

hyperparameter는 다음과 같이 설정

- epochs : 100
- hiddensize : 128, 64 (two layer)
- learning_rate : 0.5

In [5]:
from torchvision import transforms, datasets
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset
from torch.utils.data import DataLoader

In [6]:
# 이미지를 텐서로 변경
transform = transforms.Compose([
    transforms.ToTensor()
])

In [7]:
trainset = datasets.MNIST(
    root      = './.data/', 
    train     = True,
    download  = True,
    transform = transform
)
testset = datasets.MNIST(
    root      = './.data/', 
    train     = False,
    download  = True,
    transform = transform
)

  0%|          | 0/9912422 [00:00<?, ?it/s]

Extracting ./.data/MNIST/raw/train-images-idx3-ubyte.gz to ./.data/MNIST/raw



  0%|          | 0/28881 [00:00<?, ?it/s]

Extracting ./.data/MNIST/raw/train-labels-idx1-ubyte.gz to ./.data/MNIST/raw



  0%|          | 0/1648877 [00:00<?, ?it/s]

Extracting ./.data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./.data/MNIST/raw



  0%|          | 0/4542 [00:00<?, ?it/s]

Extracting ./.data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./.data/MNIST/raw



In [8]:
BATCH_SIZE = 512
# train set과 test set 각각에 대하여 DataLoader를 생성합니다.
# shuffle=True 매개변수를 넣어 데이터를 섞어주세요.
train_loader = DataLoader(trainset, batch_size=BATCH_SIZE, shuffle=True)
test_loader =  DataLoader(testset, batch_size=BATCH_SIZE, shuffle=True)

In [9]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.layer1 = nn.Linear(784,128)
        self.layer2 = nn.Linear(128,64)
        self.layer3 = nn.Linear(64,10)
        self.relu = nn.ReLU()
        
    def forward(self, x):
        x = x.view(-1, 784)
        out = self.layer1(x)
        out = self.relu(out)
        out = self.layer2(out)
        out = self.relu(out)
        out = self.layer3(out)

        return out

In [10]:
model = Net()
model

Net(
  (layer1): Linear(in_features=784, out_features=128, bias=True)
  (layer2): Linear(in_features=128, out_features=64, bias=True)
  (layer3): Linear(in_features=64, out_features=10, bias=True)
  (relu): ReLU()
)

In [11]:
list(model.parameters())

[Parameter containing:
 tensor([[-0.0316, -0.0146,  0.0147,  ...,  0.0104, -0.0261, -0.0095],
         [-0.0023,  0.0297, -0.0014,  ...,  0.0069, -0.0272, -0.0142],
         [-0.0198, -0.0213,  0.0255,  ..., -0.0353,  0.0149,  0.0050],
         ...,
         [-0.0267,  0.0236,  0.0010,  ..., -0.0120, -0.0341, -0.0117],
         [ 0.0232, -0.0347, -0.0287,  ..., -0.0296,  0.0051, -0.0335],
         [-0.0340,  0.0118, -0.0031,  ..., -0.0255,  0.0122,  0.0149]],
        requires_grad=True), Parameter containing:
 tensor([-0.0048,  0.0085, -0.0118,  0.0209,  0.0187,  0.0231,  0.0024, -0.0256,
          0.0145,  0.0247,  0.0035,  0.0339,  0.0284,  0.0148,  0.0067, -0.0186,
         -0.0034,  0.0255, -0.0190,  0.0128,  0.0264, -0.0337,  0.0092,  0.0198,
         -0.0122, -0.0243, -0.0164, -0.0029,  0.0059, -0.0084,  0.0230, -0.0177,
          0.0073, -0.0242, -0.0073, -0.0095,  0.0304, -0.0071, -0.0263, -0.0138,
         -0.0016,  0.0200,  0.0255,  0.0210,  0.0331, -0.0210,  0.0159, -0.0302,

In [12]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.5)

In [13]:
def train(model, train_loader, optimizer):
    model.train()
    # 배치 당 loss 값을 담을 리스트 생성
    batch_losses = []

    for data, target in train_loader:
        # 옵티마이저의 기울기 초기화
        optimizer.zero_grad()

        # y pred 값 산출
        output = model(data)
        # loss 계산
        # 정답 데이터와의 cross entropy loss 계산
        # 이 loss를 배치 당 loss로 보관
        loss = criterion(output, target)
        batch_losses.append(loss)

        # 기울기 계산
        loss.backward()

        # 가중치 업데이트!
        optimizer.step()
        
    # 배치당 평균 loss 계산
    avg_loss = sum(batch_losses) / len(batch_losses)
    
    return avg_loss

In [14]:
def evaluate(model, test_loader):
    # 모델을 평가 모드로 전환
    model.eval()

    batch_losses = []
    correct = 0 

    with torch.no_grad(): 
        for data, target in test_loader:
            # 예측값 생성
            output = model(data)

            # loss 계산 (이전과 동일)
            loss = criterion(output, target)
            batch_losses.append(loss)

           # Accuracy 계산
           # y pred와 y가 일치하면 correct에 1을 더해주기
            pred = output.max(1, keepdim=True)[1]

            # eq() 함수는 값이 일치하면 1을, 아니면 0을 출력.
            correct += pred.eq(target.view_as(pred)).sum().item()

    # 배치 당 평균 loss 계산 
    avg_loss =  sum(batch_losses) / len(batch_losses)

    #정확도 계산
    accuracy = 100. * correct / len(test_loader.dataset)

    return avg_loss, accuracy

In [15]:
EPOCHS = 100

for epoch in range(1, EPOCHS + 1):
    train_loss = train(model, train_loader, optimizer)
    test_loss, test_accuracy = evaluate(model, test_loader)
    
    print('[{}] Train Loss: {:.4f}\tTest Loss: {:.4f}\tAccuracy: {:.2f}%'.format(
          epoch, train_loss, test_loss, test_accuracy))

[1] Train Loss: 0.8482	Test Loss: 0.3196	Accuracy: 89.91%
[2] Train Loss: 0.2467	Test Loss: 0.2053	Accuracy: 93.66%
[3] Train Loss: 0.1717	Test Loss: 0.1790	Accuracy: 94.30%
[4] Train Loss: 0.1337	Test Loss: 0.1249	Accuracy: 96.17%
[5] Train Loss: 0.1073	Test Loss: 0.1058	Accuracy: 96.67%
[6] Train Loss: 0.0927	Test Loss: 0.0943	Accuracy: 96.97%
[7] Train Loss: 0.0794	Test Loss: 0.1072	Accuracy: 96.62%
[8] Train Loss: 0.0685	Test Loss: 0.0906	Accuracy: 96.96%
[9] Train Loss: 0.0598	Test Loss: 0.0803	Accuracy: 97.43%
[10] Train Loss: 0.0548	Test Loss: 0.0871	Accuracy: 97.44%
[11] Train Loss: 0.0469	Test Loss: 0.0900	Accuracy: 97.28%
[12] Train Loss: 0.0423	Test Loss: 0.1157	Accuracy: 96.63%
[13] Train Loss: 0.0375	Test Loss: 0.0745	Accuracy: 97.65%
[14] Train Loss: 0.0345	Test Loss: 0.1103	Accuracy: 96.66%
[15] Train Loss: 0.0308	Test Loss: 0.0848	Accuracy: 97.37%
[16] Train Loss: 0.0268	Test Loss: 0.2211	Accuracy: 93.85%
[17] Train Loss: 0.6960	Test Loss: 0.1703	Accuracy: 94.87%
[18] T

## 과제 4
과제 3 부분의 성능을 지금까지 배운 지식을 바탕으로 향상시켜보세요

- Hint : Activation function, hyperparameter setting

In [16]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.layer1 = nn.Linear(784,128)
        self.layer2 = nn.Linear(128,64)
        self.layer3 = nn.Linear(64,10)
        self.relu = nn.ReLU()
        self.softmax = nn.Softmax(dim=1) #softmax 활성화 함수 사용
        
    def forward(self, x):
        x = x.view(-1, 784)
        out = self.layer1(x)
        out = self.relu(out)
        out = self.layer2(out)
        out = self.relu(out)
        out = self.layer3(out)
        out = self.softmax(out)

        return out

In [17]:
model = Net()
model

Net(
  (layer1): Linear(in_features=784, out_features=128, bias=True)
  (layer2): Linear(in_features=128, out_features=64, bias=True)
  (layer3): Linear(in_features=64, out_features=10, bias=True)
  (relu): ReLU()
  (softmax): Softmax(dim=1)
)

In [18]:
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.5)

In [19]:
EPOCHS = 100

for epoch in range(1, EPOCHS + 1):
    train_loss = train(model, train_loader, optimizer)
    test_loss, test_accuracy = evaluate(model, test_loader)
    
    print('[{}] Train Loss: {:.4f}\tTest Loss: {:.4f}\tAccuracy: {:.2f}%'.format(
          epoch, train_loss, test_loss, test_accuracy))

[1] Train Loss: 2.2455	Test Loss: 1.9954	Accuracy: 52.88%
[2] Train Loss: 1.8449	Test Loss: 1.7952	Accuracy: 69.08%
[3] Train Loss: 1.7382	Test Loss: 1.7201	Accuracy: 74.79%
[4] Train Loss: 1.6915	Test Loss: 1.6446	Accuracy: 82.56%
[5] Train Loss: 1.6366	Test Loss: 1.6287	Accuracy: 83.87%
[6] Train Loss: 1.6244	Test Loss: 1.6265	Accuracy: 84.08%
[7] Train Loss: 1.6179	Test Loss: 1.6259	Accuracy: 83.89%
[8] Train Loss: 1.6135	Test Loss: 1.6113	Accuracy: 85.22%
[9] Train Loss: 1.6097	Test Loss: 1.6104	Accuracy: 85.26%
[10] Train Loss: 1.6069	Test Loss: 1.6142	Accuracy: 85.11%
[11] Train Loss: 1.6051	Test Loss: 1.6099	Accuracy: 85.46%
[12] Train Loss: 1.6024	Test Loss: 1.6057	Accuracy: 85.95%
[13] Train Loss: 1.6000	Test Loss: 1.6018	Accuracy: 86.09%
[14] Train Loss: 1.5978	Test Loss: 1.6009	Accuracy: 86.20%
[15] Train Loss: 1.5960	Test Loss: 1.6254	Accuracy: 83.81%
[16] Train Loss: 1.5948	Test Loss: 1.5966	Accuracy: 86.56%
[17] Train Loss: 1.5926	Test Loss: 1.5946	Accuracy: 86.85%
[18] T

**무엇을 보완하였고, 왜 보완되었는지에 대한 자유 서술 (아래에)**

활성화함수를 softmax 함수로 바꿨는데, accuracy 가 오히려 97.7% -> 97.49% 로 떨어졌다...

In [20]:
EPOCHS = 120

for epoch in range(1, EPOCHS + 1):
    train_loss = train(model, train_loader, optimizer)
    test_loss, test_accuracy = evaluate(model, test_loader)
    
    print('[{}] Train Loss: {:.4f}\tTest Loss: {:.4f}\tAccuracy: {:.2f}%'.format(
          epoch, train_loss, test_loss, test_accuracy))

[1] Train Loss: 1.4709	Test Loss: 1.4867	Accuracy: 97.49%
[2] Train Loss: 1.4709	Test Loss: 1.4865	Accuracy: 97.48%
[3] Train Loss: 1.4708	Test Loss: 1.4865	Accuracy: 97.54%
[4] Train Loss: 1.4708	Test Loss: 1.4863	Accuracy: 97.51%
[5] Train Loss: 1.4708	Test Loss: 1.4865	Accuracy: 97.53%
[6] Train Loss: 1.4707	Test Loss: 1.4867	Accuracy: 97.61%
[7] Train Loss: 1.4710	Test Loss: 1.4867	Accuracy: 97.50%
[8] Train Loss: 1.4707	Test Loss: 1.4863	Accuracy: 97.55%
[9] Train Loss: 1.4706	Test Loss: 1.4864	Accuracy: 97.57%
[10] Train Loss: 1.4705	Test Loss: 1.4867	Accuracy: 97.50%
[11] Train Loss: 1.4708	Test Loss: 1.4866	Accuracy: 97.48%
[12] Train Loss: 1.4704	Test Loss: 1.4865	Accuracy: 97.55%
[13] Train Loss: 1.4704	Test Loss: 1.4864	Accuracy: 97.55%
[14] Train Loss: 1.4704	Test Loss: 1.4863	Accuracy: 97.58%
[15] Train Loss: 1.4703	Test Loss: 1.4866	Accuracy: 97.46%
[16] Train Loss: 1.4704	Test Loss: 1.4866	Accuracy: 97.49%
[17] Train Loss: 1.4704	Test Loss: 1.4864	Accuracy: 97.49%
[18] T

EPOCH 를 120으로 늘렸더니 accuracy가 97.6% 이 되었다. 